### PR TITLE
Support Service's Cluster IP based upstream target

### PR DESCRIPTION
**Pull request by subicura**
Option for use service's Cluster IP and port, instead of endpoints.

https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/annotations.md#service-upstream

### CODE REVIEW

**Comment by gerred** [comment]
Hi @subicura, I'm taking a look at all of the backed up PRs and issues now. Once I've had a chance to look, I'll review!

**Comment by hbagdi** [comment]
ClusterIP is a virtual IP in k8s, which is handled by kube-proxy. Using ClusterIP, will result in Kong being unable to do sticky sessions or other load balancing. It's ideal that your ingress point is aware of all the upstream pods of a service and then load balances between them.

Could you help me better understand your use case here to justify this feature?

**Comment by purplehq** [comment]
@hbagdi Hello! My case is kong + kubernets + istio. Istio(https://istio.io/) supports only Cluster IP. (Service mesh handles load balance itself.)
I know this option is not normal case. But this is useful for service mesh. Thanks!

I think this links are useful.
https://github.com/kubernetes/ingress-nginx/issues/2126
https://github.com/istio/istio/issues/4840

**Comment by subicura** [comment]
@hbagdi Hello! My case is kong + kubernets + istio. Istio(https://istio.io/) supports only Cluster IP. (Service mesh handles load balance itself.)
I know this option is not normal case. But this is useful for service mesh. Thanks!

I think this links are useful.
kubernetes/ingress-nginx#2126
istio/istio#4840

File: internal/apis/configuration/v1/types.go [hunk 1]
```diff
@@ -40,11 +40,12 @@ type Route struct {
...
-	HashOnHeader string        `json:"hash_on_header"`
-	HashFallback string        `json:"hash_fallback"`
-	Healthchecks *Healthchecks `json:"healthchecks,omitempty"`
-	Slots        int           `json:"slots"`
+	HashOn          string        `json:"hash_on"`
+	HashOnHeader    string        `json:"hash_on_header"`
+	HashFallback    string        `json:"hash_fallback"`
+	Healthchecks    *Healthchecks `json:"healthchecks,omitempty"`
+	Slots           int           `json:"slots"`
+	ServiceUpstream bool          `json:"service_upstream"`
```
Comment (hbagdi) [inline internal/apis/configuration/v1/types.go]: Can we rename this to `UseClusterIP` to clearly convey the meaning of the property?

File: internal/ingress/controller/kong.go [hunk 2]
```diff
@@ -112,11 +113,44 @@ func (n *NGINXController) OnUpdate(ingressCfg *ingress.Configuration) error {
 	return nil
 }
 
+// Duplicated by github.com/kubernetes/ingress-nginx/internal/ingress/controller/controller.go#getServiceClusterEndpoint
```
Comment (hbagdi) [inline internal/ingress/controller/kong.go]: Can you remove this line and put a more meaning comment documenting the method's purpose?

File: internal/ingress/controller/kong.go [hunk 3]
```diff
@@ -112,11 +113,44 @@ func (n *NGINXController) OnUpdate(ingressCfg *ingress.Configuration) error {
...
+}
+
 // syncTargets reconciles the state between the ingress controller and
 // kong comparing the endpoints in Kubernetes and the targets in a
 // particular kong upstream. To avoid downtimes we create the new targets
 // first and then remove the killed ones.
-func syncTargets(upstream string, ingressEndpopint *ingress.Backend, client *kong.RestClient) error {
+// If useServiceUpstream set, use service's Cluster IP and port, instead of endpoints.
+// https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/annotations.md#service-upstream
+func syncTargets(upstream string, ingressEndpopint *ingress.Backend, client *kong.RestClient, useServiceUpstream bool) error {
```
Comment (hbagdi) [inline internal/ingress/controller/kong.go]: `useServiceUpstream` doesn't convey anything meaningful. Perhaps, a better name would be `useClusterIP`?

Also, there is typo in `ingressEndpopint`. Could you correct that? I understand that the typo was not introduced in this PR. It'll be acceptable if you don't correct it as well. Thanks!

Comment (hbagdi) [reply internal/ingress/controller/kong.go]: Also, could you remove the link to the ingress-nginx reference? Thanks!

File: internal/ingress/controller/kong.go [hunk 4]
```diff
@@ -905,9 +953,16 @@ func (n *NGINXController) syncUpstreams(locations []*ingress.Location, backends
...
+			if kongIngress != nil && kongIngress.Upstream != nil && kongIngress.Upstream.ServiceUpstream {
+				err := syncTargets(upstreamName, upstream, client, true)
+				if err != nil {
+					return errors.Wrap(err, "syncing targets (use service upstream)")
+				}
+			} else {
+				err := syncTargets(upstreamName, upstream, client, false)
+				if err != nil {
+					return errors.Wrap(err, "syncing targets")
+				}
```
Comment (hbagdi) [inline internal/ingress/controller/kong.go]: Could we simplify this logic a little bit?
Probably using a local boolean variable to store whether to use ClusterIP or not and then just use a single call to  `syncTargets`?

**Comment by hbagdi** [comment]
@subicura This PR can get merged in, but requires a little bit of code hygiene work.
Also, it seems like the place to put the `getServiceClusterEndpoint` seems odd but that's okay for now. We will be refactoring this part in future.

**Comment by subicura** [comment]
@hbagdi Thanks for the detailed review. I'll try improve code quality soon.

**Comment by hbagdi** [comment]
@subicura Friendly ping for an update around this.

**Comment by hbagdi** [comment]
@subicura Another friendly nudge on this.
This feature could be useful to a wider audience if we can get it out soon.

**Comment by hbagdi** [comment]
@subicura Another friendly ping on this, this has been stalling for a while.
I'm afraid I'll have to close the PR if there is no activity.

**Comment by hbagdi** [comment]
Closing this due to no activity.
Please re-open if necessary.

**Pull request closed without merging**